play-routes-toolchain/transitions.bzl

```diff
@@ -0,0 +1,62 @@
+# This approach of incoming transition, store original values, outgoing
+# transition, reset to original values is inspired by what the rules_go
+# folks are doing.
+toolchain_setting_key = "@rules_play_routes_toolchain//:play-routes-toolchain"
+original_toolchain_setting_key = "@rules_play_routes_toolchain//:original-play-routes-toolchain"
+
+def _play_routes_toolchain_transition_impl(settings, attr):
+    """Update the rules_play_routes toolchain to the overridden value. Store
+    The original value, so it can be reset on the outgoing transition.
+    """
+    if attr.play_routes_toolchain_name == "":
+        return {}
+    elif settings[toolchain_setting_key] == attr.play_routes_toolchain_name:
+        # Do nothing when the attribute value is what the setting currently is
+        # otherwise needlessly change build settings by modifying the setting
+        # we store the original value in
+        return {}
+
+    new_settings = {}
+    new_settings[toolchain_setting_key] = attr.play_routes_toolchain_name
+
+    # Store the original toolchain value, so we can reset it
+    new_settings[original_toolchain_setting_key] = json.encode(settings[toolchain_setting_key])
+
+    return new_settings
+
+play_routes_toolchain_transition = transition(
+    implementation = _play_routes_toolchain_transition_impl,
+    inputs = [
+        toolchain_setting_key,
+        original_toolchain_setting_key,
+    ],
+    outputs = [
+        toolchain_setting_key,
+        original_toolchain_setting_key,
+    ],
+)
+
+def _reset_play_routes_toolchain_transition_impl(settings, attr):
+    """Sets the rules_play_routes toolchain to the values it had before the
+    last play_routes_toolchain_transition.
+    """
+
+    if settings[original_toolchain_setting_key] != "":
```

Comment:
Is it possible to end up in a state in which it wouldn't be equivalent? If not, then we can use the default value. 

I was defensively programming here. Are there downsides to doing it this way? If not, then I say we keep it this way to be safe. If so, then I say we go default to avoid the downsides.

Reply:
Actually, now that I think about it, you could supply a command-line argument to override the default value, in which case they wouldn't be equivalent. This makes the most sense to me.